BOLT 2: disallow sending multiple `shutdown` msg (#977)
lightning/bolts: 02-peer-protocol.md

  - MAY send a `shutdown` before a `funding_locked`, i.e. before the funding transaction has reached `minimum_depth`.
  - if there are updates pending on the receiving node's commitment transaction:
    - MUST NOT send a `shutdown`.

>   - MIST not send multiple `shutdown` message.

Reviewer: nit:
```suggestion
  - MUST not send multiple `shutdown` messages.
```

Author: Thanks to catch this!

        (witness program versions 1 through 16)

A receiving node:
  - if it receives multiple `shutdown` messages:

>     - SHOULD ignore all except the first one.

Reviewer: I agree, I hesitated to suggest removing it as well.
That would be totally fine by me (and less bloat in the spec is always a win).

>     - SHOULD ignore all except the first one.

Reviewer: > Mh! I don't like the idea to ignore messages here too because I'm not sure if we can check if a message arrives first or second right?

What do you mean? Of course you know what message arrived first as a receiver...? And since senders must not do it anyway, this can only happen for a buggy sender so you don't have to care much about the outcome, it's very likely that this will end up in a force-close anyway.

> we miss the possibility to catch this bug during the developing cycle without panic node in production

I think you're exaggerating the importance of this issue :slightly_smiling_face:

>     - SHOULD ignore all except the first one.

Author: >I think you're exaggerating the importance of this issue

Yeah, I know, but I'm just trying to find a reason to answer myself if all agree that this is a force close why not put it in the spec? I'm not trying to push in the spec repo, but I would like to know why. Just because I don't like to leave myself with pending questions.

>     - SHOULD ignore all except the first one.

Reviewer: We don't all agree it must be a force-close, there are multiple behaviors in that case that are ok.

Eclair just ignores the other `shutdown` and will try doing a mutual close based on the script received in the first `shutdown`, which may end up in a force-close if the buggy sender changed the script (but may work otherwise).

So I don't want the spec to mandate that I should force-close here, but depending on the codebase, it could make sense for others to always force-close, so I don't want to mandate anything either to them. We must give some leeway to implementations on things that don't matter much such as this to avoid over-constraining them and giving them enough room to use different architectures.

>     - SHOULD ignore all except the first one.

Reviewer 2: Comment from spec meeting: should change to say you can only send one `shutdown` per reconnection. 

The receiver can do w/e they want if this is violated technically.

Reviewer: @vincenzopalazzo if you remove these receiver requirements, I believe we can merge this as discussed during the last spec meeting?

Author: Yes! I will refine this PR by the end of the week

>     - SHOULD ignore all except the first one.

Author: This make sense, thanks @t-bast

>   - if it have received multiple `shutdown`:
>     - should ignore the others after the first one.

Reviewer: nit:
```suggestion
  - if it receives multiple `shutdown` messages:
    - SHOULD ignore all except the first one.
```

Author: Thanks! maybe before opening a PR a self-review was needed! Sorry about that
